File: sdk/src/bytes.ts
import * as bs58 from "bs58";
import { sha256 } from "@noble/hashes/sha256";
import { ed25519 } from "@noble/curves/ed25519";

export function encodePk(bytes: Uint8Array): string {
  return bs58.encode(bytes);
}

export function decodePk(s: string): Buffer {
  return Buffer.from(bs58.decode(s));
}

function isOnCurve(p: Uint8Array): boolean {
  try {
    ed25519.ExtendedPoint.fromHex(Buffer.from(p).toString("hex"));
    return true;
  } catch {
    return false;
  }
}

const PDA_MARKER = Buffer.from("ProgramDerivedAddress");

export function findProgramAddress(seeds: Uint8Array[], programId: Uint8Array): { address: string; bump: number } {
  const pid = Buffer.from(programId);
  for (let bump = 255; bump >= 0; bump--) {
    const parts = [...seeds.map((s) => Buffer.from(s)), Buffer.from([bump]), pid, PDA_MARKER];
    const hash = sha256(Buffer.concat(parts));
    if (!isOnCurve(hash)) return { address: encodePk(hash), bump };
  }
  throw new Error("unable to find PDA");
}

File: sdk-rs/src/bytes.rs
use curve25519_dalek::edwards::CompressedEdwardsY;
use sha2::{Digest, Sha256};

const PDA_MARKER: &[u8] = b"ProgramDerivedAddress";

pub fn encode_pk(bytes: &[u8]) -> String {
    bs58::encode(bytes).into_string()
}

pub fn decode_pk(s: &str) -> Result<[u8; 32], String> {
    let v = bs58::decode(s).into_vec().map_err(|e| e.to_string())?;
    v.try_into().map_err(|_| "pubkey must be 32 bytes".into())
}

fn on_curve(p: &[u8; 32]) -> bool {
    CompressedEdwardsY(*p).decompress().is_some()
}

pub fn find_program_address(seeds: &[&[u8]], program_id: &[u8; 32]) -> Result<(String, u8), String> {
    for bump in (0u8..=255).rev() {
        let mut h = Sha256::new();
        for s in seeds {
            h.update(s);
        }
        h.update([bump]);
        h.update(program_id);
        h.update(PDA_MARKER);
        let hash: [u8; 32] = h.finalize().into();
        if !on_curve(&hash) {
            return Ok((encode_pk(&hash), bump));
        }
    }
    Err("unable to find PDA".into())
}
